File: frontend-dashboard/lib/api/dm-charts.ts
import { z } from "zod";
import { http } from "./client";
import { WarmProgressSchema } from "./warm-progress";

// Bar-1 population is "DM murni" (glukosa mencapai ambang diagnosis: GDP>=126 /
// GD2PP>=200 / GDS-2>=200), NOT registry membership (which also admits
// Prediabetes + riwayat-gated-but-normal patients) — see backend
// dm_charts_scan._dm_murni. The 2026 Ya/Tidak split uses the same diagnosis
// threshold (symmetric with the hipertensi charts).
export const DmKohort2TahunSchema = z.object({
  dm_2025: z.number().int().default(0),
  diperiksa_2026: z.number().int().default(0),
  dm_2026_tinggi: z.number().int().default(0),
  dm_2026_terkendali: z.number().int().default(0),
  // Each split from dm_2026_tinggi / dm_2026_terkendali, not a total.
  tinggi_diobati: z.number().int().default(0),
  tinggi_tidak_diobati: z.number().int().default(0),
  terkendali_diobati: z.number().int().default(0),
  terkendali_tidak_diobati: z.number().int().default(0),
});

export const DmCharts2026Schema = z.object({
  dm_2026: z.number().int().default(0),
  pasien_baru: z.number().int().default(0),
  sudah_dm: z.number().int().default(0),
  // Each split from pasien_baru / sudah_dm, not a total.
  baru_diobati: z.number().int().default(0),
  baru_tidak_diobati: z.number().int().default(0),
  sudah_diobati: z.number().int().default(0),
  sudah_tidak_diobati: z.number().int().default(0),
});

export const DmChartsSchema = z.object({
  kohort_2tahun: DmKohort2TahunSchema,
  dm_2026: DmCharts2026Schema,
  cache_hit: z.boolean(),
  computed_at: z.string().nullable(),
  computing: z.boolean().default(false),
  progress: WarmProgressSchema.nullable().default(null),
});
export type DmCharts = z.infer<typeof DmChartsSchema>;

export async function getDmCharts(puskesmas_id: string): Promise<DmCharts> {
  const { data } = await http.get(`/dm-reports/charts`, {
    params: { puskesmas_id },
  });
  return DmChartsSchema.parse(data);
}

File: backend/app/schemas/dm_charts.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.report_progress import WarmProgress


class DmKohort2Tahun(BaseModel):
    """Chart 1 — DM 2 tahun berturut-turut (absolute counts). Bar-1 population is
    "DM murni" (glukosa mencapai ambang diagnosis), NOT registry membership —
    see ``dm_charts_scan._dm_murni``."""

    dm_2025: int = 0
    diperiksa_2026: int = 0
    dm_2026_tinggi: int = 0  # glukosa 2026 masih >= ambang diagnosis DM
    dm_2026_terkendali: int = 0  # glukosa 2026 di bawah ambang (dari diperiksa_2026)
    # Diobati/tidak, masing-masing dipecah dari dm_2026_tinggi / dm_2026_terkendali
    # (bukan total dari diperiksa_2026) — pernah diobati kapan saja 2025-2026.
    tinggi_diobati: int = 0
    tinggi_tidak_diobati: int = 0
    terkendali_diobati: int = 0
    terkendali_tidak_diobati: int = 0


class DmCharts2026(BaseModel):
    """Chart 2 — DM CKG 2026 (absolute counts). Bar-1 population is "DM murni"
    (glukosa mencapai ambang diagnosis), NOT registry membership."""

    dm_2026: int = 0
    pasien_baru: int = 0  # tanpa riwayat DM
    sudah_dm: int = 0  # ada riwayat DM
    # Diobati/tidak, masing-masing dipecah dari pasien_baru / sudah_dm
    # (bukan total dari dm_2026) — pernah diobati kapan saja 2025-2026.
    baru_diobati: int = 0
    baru_tidak_diobati: int = 0
    sudah_diobati: int = 0
    sudah_tidak_diobati: int = 0


class DmChartsOut(BaseModel):
    """The 2-chart DM dashboard payload for one puskesmas (rolling cross-year
    window). Carries ABSOLUTE counts; percentages are computed on the frontend.
    Same cache-metadata triad as the registry dashboards, so a cold cache returns
    immediately and the caller polls back."""

    kohort_2tahun: DmKohort2Tahun = Field(default_factory=DmKohort2Tahun)
    dm_2026: DmCharts2026 = Field(default_factory=DmCharts2026)
    cache_hit: bool
    computed_at: datetime | None
    computing: bool = False
    progress: WarmProgress | None = None  # approx warm progress while computing
